### ISSUE TITLE
Monorepo Migration Mega-issue

### ISSUE THREAD

**Issue by pjhul** [PostHog/apps#1]
### Welcome to the (future) home of all PostHog apps :partying_face: 

I'm creating this issue to give an overview of the migration process I've put together so far and hope this will give everyone a good idea of what I'll be working on in the coming weeks. The current plan (at a high level) is to publish separate NPM packages for each app, and then create a migration to update all the official plugins on Cloud and Self-host to point to these packages.

This will provide a number of benefits:
- Better observability on issues as they will all be centralized
- Helper functions can be shared between apps
- New apps can be reviewed/added through PR instead of forking
- Full E2E testing
- Ability to use 3rd-party NPM modules

### Preliminary
- [x] Sync existing apps
- [x] Create `manifest.json` to mirror [PostHog/integrations-repository](https://github.com/PostHog/integrations-repository) with all the apps contained in the monorepo
- [x] Setup publishing to NPM
- [x] https://github.com/PostHog/posthog/pull/12356 _(mainly for debugging)_

### Switch-over
- [x] https://github.com/PostHog/posthog/pull/12438
- [ ] Update existing app `README.md`'s with note that eventually they will no longer be maintained and all issues should be filed here
- [ ] Update docs on submitting a new app
- [ ] Update `plugin.json` fetching on posthog.com
- [ ] #8 
- [ ] Sync existing issues from each repo

### Future work
- [ ] Full E2E testing
- [ ] Module bundling and breaking shared code into a central library of helpers

This is still fairly early on and there's a lot of fleshing-out I still need to do, but I've done enough digging to where I'm happy with this plan at a high level. As always, if there's anything I've missed or any feedback you have on the overall process please don't hesitate to add/edit things! 

cc: @yakkomajuri @mariusandra @tiina303 @macobo @timgl @jamesefhawkins

**Comment by mariusandra** [comment]
Some thoughts:
- I'm not sure what the "dist" folder is for. Apps don't need to be transpiled... and it's even better if we explicitly don't transpile them, and always load the original TypeScript source in the plugin server. That's because the plugin server injects various safeguards into the plugins (infinite loop busters, etc)... which we want to have control over.
- Publishing into NPM is one way of getting the apps up, that I just listed as an alternative once, without saying this is what you should go for. I still haven't had the time to review the app subfolders PR (sorry 😅), but if that works, I'm not sure if there's a strong need to use NPM. I'm not against it though, just clarifying my stance...
- There's one more thing that's different for NPM vs GitHub -> when upgrading, if it's a GitHub package, we can show a link to the diff between the commits right there. This way you always know what changed when upgrading. Unfortunately you can't see a diff like that between NPM package versions. The most we could do is to link back to the repo with a version/tag range, but that won't work since git tags are global and not meant to tag multiple different packages in a repo... 🤔

**Comment by pjhul** [comment]
@mariusandra Thanks a ton for the thoughts!

> * I'm not sure what the "dist" folder ...

Would these safeguards be broken by this, even without minification and targeting the latest Node features?. Haven't looked into this area though so don't really haven much context on these.

The main reason I opted for this wasn't strictly transpilation and was more for bundling - it would be _super_ nice to be able to use third-party NPM packages without having to explicitly allow and include them on the plugin server side. Along with this, having a shared library of helper functions that all plugins can access would be amazing. 

If this does break those safeguards I'm fine opting for just publishing the source files directly, but from a DX perspective I would _love_ to have this.

> * Publishing into NPM is one way of getting the apps up, that I just listed as an alternative once, ...

Totally agree, and while I don't think there's a super strong reason for one or the other (I went back and forth _a lot_ :sweat_smile:), I eventually settled on the NPM route mainly because it felt much easier to manage releases. By using semver's instead of commits it makes it much easier to do things like: show what type of update is available for an app (patch, major, etc) as well as displaying a human-readable version to the user. As things grow, I also like that we can chunk multiple commits into a release rather than every single commit to `main` counting as a new version.

> * There's one more thing that's different for NPM vs GitHub -> when upgrading, ...

I'm not _too_ worried about losing this, as I would assume that not _a ton_ of users are checking the raw diffs, but I could be totally wrong (not sure if we track this?). Furthermore, I think using the version to display whether the update is a major/patch/etc would give the user enough context.

Another thing that might be interesting is somehow publishing a changelog showing all the commits in a release. Haven't done any thinking about how this would actually work, but I'd imagine most users would prefer this to a raw diff :shrug:

**Comment by mariusandra** [comment]
> Would these safeguards be broken by this, even without minification and targeting the latest Node features?. Haven't looked into this area though so don't really haven much context on these.

We'd still re-transpile the compiled source in `dist` as it is now, and perhaps double add some safeguards. We really want to make sure that you can't take the server down by just writing `function processEvent() { for(;;); }`. Since we can't trust code that comes from any external source, and we also want to protect users from themselves (accidental endless loop by missing a symbol?), we have no other option than to re-transpile the code and add our safeguards.

The way I look at it is that with the plugin server we're basically building a custom JS runtime (actually just wrapping node), with the raw apps themselves being the input for this runtime. Imagine running `plugin-server customerio-app/index.ts` the same way you'd run `deno customerio-app/index.ts`.

> The main reason I opted for this wasn't strictly transpilation and was more for bundling - it would be super nice to be able to use third-party NPM packages without having to explicitly allow and include them on the plugin server side. Along with this, having a shared library of helper functions that all plugins can access would be amazing.

I think that's a solved problem, just not implemented. :) 

The answer is import maps and CDNs like `jspm.org`. Take [this generated import map](https://generator.jspm.io/#02RhYGBiDs0rySzJSU1hyE5NdDDWM9AzYShId7DQs9AzAADeSx3mIQA).

<img width="1457" alt="image" src="https://user-images.githubusercontent.com/53387/198390394-caded3fb-9047-442f-aebb-c83e8e920416.png">

Now each dependency is just a few JS files that you download, cache and serve with the app in addition to the [global imports object](https://github.com/PostHog/posthog/blob/9c6f20b697b1db40b7749aa5b46e25b22d931955/plugin-server/src/worker/vm/imports.ts#L21).

It's a bit more complicated than that, but definitely doable.

> I'm not too worried about losing this, as I would assume that not a ton of users are checking the raw diffs, but I could be totally wrong (not sure if we track this?). Furthermore, I think using the version to display whether the update is a major/patch/etc would give the user enough context.

I use it every time I upgrade an app on cloud to be absolutely sure nothing weird is ending up in production. If I wouldn't have this, I'd probably open the repo manually and check the last commits. Making this a link to a changelog (if set in plugin.json?) or just a link to the the commit history for a github folder will probably be enough.

The danger with npm packages though is that you can publish anything on npm. It doesn't have to be the same code that's on github. Thus to be really sure I'd still somehow like to compare the contents of the packages. However we might even need to build our own diffing tool inside posthog for this. We'd just be diffing two tarballs, and showing users the raw output, similar to the command line `npm diff --diff=posthog-js@1.33.0 --diff=posthog-js@1.34.0`. There are [existing online npm diffing tools](https://npm-diff.app/posthog-js-lite@0.0.5...posthog-js-lite@0.0.4), though I don't know if we could link to them, or are they now a "data processor" that we need to handle?

All solvable problems in any case.

**Comment by mariusandra** [comment]
Regarding the `dist` folder, I still suggest just getting rid of it. You wouldn't commit it to github for a npm package, and it's just another step in another pipeline that can cause problems. 

The other thing I'd do is to rename `src/packages` to just `packages`. There's no need for the extra nesting, and if any package needs a separate `src` folder, they're free to add it.

**Comment by pjhul** [comment]
Fair enough, for this first version I'll go ahead and just publish the source directly. I'll keep the rollup stuff on a separate branch to play around with in the future, but for now I agree that keeping things simple is important :raised_hands:

**Comment by lharries** [comment]
Good work @pauldambra! What was the reason for apps being a separate repo vs being a folder within posthog? Not saying one is right or wrong just keen to understand the rationale

**Comment by mariusandra** [comment]
wrong paul @lharries :D

I think this will be much easier to manage and maintain in a separate repo. Otherwise these apps don't feel like separate apps, but part of posthog-core, which is a feeling we'd like to avoid.

Plus, there's a technical reason: if you want to install an app from a folder, you need to download the whole repository as a zip and extract that one folder. You can't just download that one folder as a zip. When we install apos, this is what we do. We also then store the zip inside postgres. 

Having a separate apps folder would allow us, if needed to install custom apps from custom folders on custom branches and play around. Having it all on the posthog.com repo would make this immensely slow, as you'd have to download a LOT of data just to get one JS file.

**Comment by lharries** [comment]
Understood, thanks Marius!

**Comment by benjackwhite** [comment]
Just saw this and it looks super cool!
